compositor_render/src/transformations/builtin/node.rs

```diff
@@ -1,101 +1,54 @@
-use std::time::Duration;
+use std::{sync::Arc, time::Duration};
 
 use compositor_common::{
     renderer_spec::FallbackStrategy,
-    scene::{builtin_transformations::BuiltinSpec, NodeId, Resolution},
+    scene::{
+        builtin_transformations::{BuiltinSpec, TransformToResolutionStrategy},
+        NodeId, Resolution,
+    },
 };
 
-use wgpu::util::DeviceExt;
-
-use crate::renderer::{texture::NodeTexture, RenderCtx, WgpuCtx};
-
-use super::{params::BuiltinParams, Builtin};
-
-struct ParamsBuffer {
-    bind_group: wgpu::BindGroup,
-    buffer: wgpu::Buffer,
-    content: bytes::Bytes,
-}
-
-impl ParamsBuffer {
-    pub fn new(content: bytes::Bytes, wgpu_ctx: &WgpuCtx) -> Self {
-        let content_or_zero = match content.is_empty() {
-            true => bytes::Bytes::copy_from_slice(&[0]),
-            false => content.clone(),
-        };
-
-        let buffer = wgpu_ctx
-            .device
-            .create_buffer_init(&wgpu::util::BufferInitDescriptor {
-                label: Some("builtin node params buffer"),
-                usage: wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST,
-                contents: &content_or_zero,
-            });
-
-        let bind_group = wgpu_ctx
-            .device
-            .create_bind_group(&wgpu::BindGroupDescriptor {
-                label: Some("builtin node params bind group"),
-                layout: &wgpu_ctx.shader_parameters_bind_group_layout,
-                entries: &[wgpu::BindGroupEntry {
-                    binding: 0,
-                    resource: buffer.as_entire_binding(),
-                }],
-            });
-
-        Self {
-            bind_group,
-            buffer,
-            content,
-        }
-    }
-
-    pub fn update(&mut self, content: bytes::Bytes, wgpu_ctx: &WgpuCtx) {
-        if self.content.len() != content.len() {
-            *self = Self::new(content, wgpu_ctx);
-            return;
-        }
+use crate::{
+    gpu_shader::GpuShader,
+    renderer::{texture::NodeTexture, RenderCtx},
+    utils::rgba_to_wgpu_color,
+};
 
-        if self.content != content {
-            wgpu_ctx.queue.write_buffer(&self.buffer, 0, &content);
-        }
-    }
-}
+use super::{params::RenderParams, shader_params::ParamsBuffer, BuiltinState};
 
 pub struct BuiltinNode {
-    builtin: Builtin,
+    state: BuiltinState,
```

Comment:
[Important]
I don't like the idea of putting this here. Some builtin transformations will require having their own `state` (e.g. `tailed layout` to smoothly change inputs), so putting here `BuiltinState`, which implements `transition` logic, seems wrong to me.

I would prefer to separate `builtins` from `transitions`, such that removing sth from / adding sth to `transitions` wouldn't require any changes here. Builtin could have an `update params` function or sth like that called from `TransitionNode` before `render` and this wouldn't need to be here. Having it here also complicates every method in `Builtin`